onnxruntime/python/onnxruntime_pybind_mlvalue.cc

```diff
@@ -214,25 +213,33 @@ AllocatorPtr GetDmlAllocator(OrtDevice::DeviceId id) {
   if (hit == id_to_allocator_map->end()) {
     constexpr uint32_t device_id = 0;
     auto d3d12_device = onnxruntime::DMLProviderFactoryCreator::CreateD3D12Device(device_id, false);
-    auto dml_device = onnxruntime::DMLProviderFactoryCreator::CreateDMLDevice(d3d12_device.Get());
 
-    D3D12_COMMAND_QUEUE_DESC cmd_queue_desc = {};
-    cmd_queue_desc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
-    cmd_queue_desc.Flags = D3D12_COMMAND_QUEUE_FLAG_DISABLE_GPU_TIMEOUT;
+    ComPtr<Dml::ExecutionContext> context;
+    uint32_t execution_context_ptr_size = gsl::narrow_cast<uint32_t>(sizeof(context.GetAddressOf()));
 
-    ComPtr<ID3D12CommandQueue> cmd_queue;
-    ORT_THROW_IF_FAILED(
-        d3d12_device->CreateCommandQueue(&cmd_queue_desc, IID_PPV_ARGS(cmd_queue.ReleaseAndGetAddressOf())));
+    // First, check if an I/O binding API that was used before this session or another session has already created a queue
+    if (FAILED(d3d12_device->GetPrivateData(dml_execution_context_guid, &execution_context_ptr_size, context.GetAddressOf()))) {
+      D3D12_COMMAND_QUEUE_DESC cmd_queue_desc = {};
+      cmd_queue_desc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+      cmd_queue_desc.Flags = D3D12_COMMAND_QUEUE_FLAG_DISABLE_GPU_TIMEOUT;
 
-    auto context = std::make_shared<Dml::ExecutionContext>(d3d12_device.Get(), dml_device.Get(), cmd_queue.Get());
+      ComPtr<ID3D12CommandQueue> cmd_queue;
+      ORT_THROW_IF_FAILED(d3d12_device->CreateCommandQueue(&cmd_queue_desc, IID_PPV_ARGS(cmd_queue.ReleaseAndGetAddressOf())));
 
-    // We leak the upload and readback heaps to keep them alive, just like the map
-    auto upload_heap = std::make_unique<Dml::PooledUploadHeap>(d3d12_device.Get(), context).release();
-    auto readback_heap = std::make_unique<Dml::ReadbackHeap>(d3d12_device.Get(), context).release();
+      auto dml_device = onnxruntime::DMLProviderFactoryCreator::CreateDMLDevice(d3d12_device.Get());
+      ORT_THROW_IF_FAILED(d3d12_device->SetPrivateDataInterface(dml_device_guid, dml_device.Get()));
+
+      context = wil::MakeOrThrow<Dml::ExecutionContext>(d3d12_device.Get(), dml_device.Get(), cmd_queue.Get(), true, true);
+      ORT_THROW_IF_FAILED(d3d12_device->SetPrivateDataInterface(dml_execution_context_guid, context.Get()));
+    }
+
+    // We leak the readback and upload heap to keep them alive, just like the map
```

Comment:
:( It may be challenging, but is there a way to clean up properly?

Reply:
We could make them `ComPtr` and add them to the device (so they would be cleaned up automatically when nothing holds the device anymore). For now I'll keep it that way though since it will require changing dozens of files.